compiler/luci/pass/src/ResolveCustomOpMaxPoolWithArgmaxPass.cpp

```diff
@@ -0,0 +1,774 @@
+luci::CircleConst *create_scalar(loco::Graph *graph, Numeric value)
+{
+  auto scalar = graph->nodes()->create<luci::CircleConst>();
+
+  scalar->dtype(DT);
+
+  scalar->rank(0);
+  scalar->size<DT>(1);
+  scalar->shape_status(luci::ShapeStatus::VALID);
+
+  scalar->scalar<DT>() = value;
+
+  return scalar;
+}
+
+template <loco::DataType DT>
+void fill_coords_addition(luci::Padding padding, const luci::Stride &stride,
+                          const luci::Filter &filter, uint32_t, uint32_t input_width,
+                          luci::CircleConst *cords)
+{
+  auto const output_height = static_cast<int32_t>(cords->dim(1).value());
+  auto const output_width = static_cast<int32_t>(cords->dim(2).value());
+
+  {
+    auto const element_counts = 1 * output_height * output_width * 1;
+    cords->size<DT>(element_counts);
+  }
+
+  assert(padding != luci::Padding::UNDEFINED);
+
+  // For VALID padding:
+  int32_t start_y = 0;
+  int32_t start_x = 0;
+
+  // For SAME padding:
+  if (padding == luci::Padding::SAME)
+  {
+    start_y -= static_cast<int32_t>(filter.h() - 1) / 2;
+    start_x -= static_cast<int32_t>(filter.w() - 1) / 2;
+  }
+
+  auto const step_y = static_cast<int32_t>(stride.h());
+  auto const step_x = static_cast<int32_t>(stride.w());
+
+  for (int32_t y_o = 0, y_i = start_y; y_o < output_height; ++y_o, y_i += step_y)
+  {
+    for (int32_t x_o = 0, x_i = start_y; x_o < output_width; ++x_o, x_i += step_x)
+    {
+      auto const output_idx = y_o * output_width + x_o;
+      auto const input_idx = y_i * input_width + x_i;
+
+      cords->at<DT>(output_idx) = input_idx;
+    }
+  }
+}
+
+luci::CircleConst *create_coords_addition(luci::Padding padding, const luci::Stride &stride,
+                                          const luci::Filter &filter, uint32_t input_height,
+                                          uint32_t input_width, luci::CircleNode *pooled)
+{
+  assert(pooled->rank() == 4 && pooled->dim(3) == 1 && pooled->dim(0) == 1);
+
+  auto cords = pooled->graph()->nodes()->create<luci::CircleConst>();
+
+  cords->dtype(pooled->dtype());
+  copy_shape(pooled, cords);
+
+  switch (cords->dtype())
+  {
+    case loco::DataType::FLOAT32:
+      fill_coords_addition<loco::DataType::FLOAT32>(padding, stride, filter, input_height,
+                                                    input_width, cords);
+      break;
+    case loco::DataType::U8:
+      fill_coords_addition<loco::DataType::U8>(padding, stride, filter, input_height, input_width,
+                                               cords);
+      break;
+    case loco::DataType::S16:
+      fill_coords_addition<loco::DataType::S16>(padding, stride, filter, input_height, input_width,
+                                                cords);
+      break;
+    default:
+      // TODO support other data types
+      INTERNAL_EXN("Datatype is not supported, yet!");
+  }
+
+  return cords;
+}
+
+bool resolve_max_pool_with_argmax(luci::CircleCustom *cop)
```

Comment:
this method is very long.. it would be better to split to several functions.

Reply:
I will try to make this function shorter =)

Reply:
@Bronnikoff
You can make functions for operations and leave only high level structure.
Something like this:
``` c++
luci::CircleMaxPool2D *max_pool = make_maxpool(input);
std::vector<luci::CircleSplitOut> channel_split = make_channel_split(input);
for (int i = 0 ; i<input_depth; ++i)
{
    luci::CircleConv2D *conv = make_conv(channel_split[i]);
    etc...
}
luci::CircleConcat channels_concat = make_concat(argmax_results);
```